digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/linearasset/NumericalLimitFiller.scala

```diff
@@ -162,6 +162,171 @@ object NumericalLimitFiller {
+        case (Some(x),None) => Seq(SegmentPiece(x.id, startM, endM, SideCode.AgainstDigitizing, x.value))
+        case (None,Some(y)) => Seq(SegmentPiece(y.id, startM, endM, SideCode.TowardsDigitizing, y.value))
+        case _ => Seq()
+      }
+    }
+    /**
+      * Combine sides together if matching m-values and segment values.
+      * @param segmentPieces Sequence of one or two segment pieces (guaranteed by squash operation)
+      * @param segments All segments on road link
+      * @return Sequence of segment pieces (1 or 2 segment pieces in sequence)
+      */
+    def combineEqualValues(segmentPieces: Seq[SegmentPiece], segments: Seq[PersistedLinearAsset]): Seq[SegmentPiece] = {
+      if (segmentPieces.size < 2)
+        return segmentPieces
+
+      val seg1 = segmentPieces.head
+      val seg2 = segmentPieces.last
+      (seg1.value, seg2.value) match {
+        case (Some(v1), Some(v2)) =>
+          val sl1 = segments.find(_.id == seg1.assetId).get
+          val sl2 = segments.find(_.id == seg2.assetId).get
+          if (v1.equals(v2) && sl1.startMeasure.equals(sl2.startMeasure) && sl1.endMeasure.equals(sl2.endMeasure)) {
+            val winner = segments.filter(l => l.id == seg1.assetId || l.id == seg2.assetId).sortBy(s =>
+              s.endMeasure - s.startMeasure).head
+            Seq(segmentPieces.head.copy(assetId = winner.id, sideCode = SideCode.BothDirections))
+          } else {
+            segmentPieces
+          }
+        case (Some(v1), None) => Seq(segmentPieces.head.copy(sideCode = SideCode.BothDirections))
+        case (None, Some(v2)) => Seq(segmentPieces.last.copy(sideCode = SideCode.BothDirections))
+        case _ => segmentPieces
+      }
+    }
+
+    /**
+      * Take pieces of segment and combine two of them if they are related to the similar segment and extend each other
+      * (matching startM, endM, value and sidecode)
+      * Return altered PersistedLinearAsset as well as the pieces that could not be added to that segments.
+      * @param segmentPieces
+      * @param segments
+      * @return
+      */
+    def extendOrDivide(segmentPieces: Seq[SegmentPiece], segments: PersistedLinearAsset): (PersistedLinearAsset, Seq[SegmentPiece]) = {
+      val sorted = segmentPieces.sortBy(_.startM)
+      val current = sorted.head
+      val rest = sorted.tail
+      if (rest.nonEmpty) {
+        val next = rest.find(sp => sp.startM == current.endM && sp.sideCode == current.sideCode &&
+          sp.value == current.value)
+        if (next.nonEmpty) {
+          return extendOrDivide(Seq(current.copy(endM = next.get.endM)) ++ rest.filterNot(sp => sp.equals(next.get)), segments)
+        }
+      }
+      (segments.copy(sideCode = current.sideCode.value, startMeasure = current.startM, endMeasure = current.endM), rest)
+    }
+
+    /**
+      * Creates speed limits from orphaned segments (segments originating from a speed limit but no longer connected
```

Comment:
Maybe was copied from the SpeedLimits classes, just modify the comment and it's fine.